cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.data.stream.service;
+
+import co.cask.cdap.api.data.stream.StreamSpecification;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.conf.PropertyStore;
+import co.cask.cdap.common.io.Codec;
+import co.cask.cdap.common.zookeeper.coordination.BalancedAssignmentStrategy;
+import co.cask.cdap.common.zookeeper.coordination.PartitionReplica;
+import co.cask.cdap.common.zookeeper.coordination.ResourceCoordinator;
+import co.cask.cdap.common.zookeeper.coordination.ResourceCoordinatorClient;
+import co.cask.cdap.common.zookeeper.coordination.ResourceHandler;
+import co.cask.cdap.common.zookeeper.coordination.ResourceModifier;
+import co.cask.cdap.common.zookeeper.coordination.ResourceRequirement;
+import co.cask.cdap.common.zookeeper.store.ZKPropertyStore;
+import co.cask.cdap.data2.transaction.stream.StreamAdmin;
+import com.google.common.base.Functions;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import org.apache.twill.api.ElectionHandler;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.internal.zookeeper.LeaderElection;
+import org.apache.twill.zookeeper.ZKClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link StreamCoordinator} uses ZooKeeper to implementation coordination needed for stream. It also uses a
+ * {@link ResourceCoordinator} to elect each handler as the leader of a set of streams.
+ */
+@Singleton
+public final class DistributedStreamCoordinator extends AbstractStreamCoordinator {
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedStreamCoordinator.class);
+
+  public static final String STREAMS_COORDINATOR = "streams.coordinator";
+  private static final String STREAMS_RESOURCE = "streams";
```

Comment:
You should use `Constants.Service.STREAMS` as the resource name always need to match with the discoverable service name.